webui/react/src/stores/projects.tsx

```diff
@@ -1,156 +1,51 @@
 import { Map } from 'immutable';
-import React, {
-  createContext,
-  PropsWithChildren,
-  useCallback,
-  useContext,
-  useEffect,
-  useState,
-} from 'react';
 
 import { getWorkspaceProjects } from 'services/api';
 import { Project } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
-import { observable, useObservable, WritableObservable } from 'utils/observable';
+import { Observable, observable, WritableObservable } from 'utils/observable';
+
+class ProjectStore {
+  #projects: WritableObservable<Map<number, Project>> = observable(Map());
+  #projectsByWorkspace: WritableObservable<Map<string, Project[]>> = observable(Map());
+
+  public fetch(workspaceId: number, signal?: AbortSignal): () => void {
+    const canceler = new AbortController();
+
+    getWorkspaceProjects({ id: workspaceId, limit: 0 }, { signal: signal ?? canceler.signal })
+      .then((response) => {
+        this.#projects.update((prev) =>
+          prev.withMutations((map) => {
+            response.projects.forEach((project) => map.set(project.id, project));
+          }),
+        );
+        this.#projectsByWorkspace.update((prev) =>
+          prev.set(workspaceId.toString(), response.projects),
+        );
+      })
+      .catch(handleError);
 
-class ProjectService {
-  private _projects: WritableObservable<Map<number, Project>> = observable(Map());
-  private _projectsByIndex: WritableObservable<Map<string, Project[]>> = observable(Map());
-
-  genWorkspaceKey = (workspaceId: number): string => `byworkspace-${workspaceId}`;
-
-  fetchWorkspaceProjects = async (
-    workspaceId: number,
-    canceler: AbortController,
-    forceFetch = false,
```

Comment:
The `forceFetch` functionality is very important to the performance of the page. We shouldn't be removing necessary functionality in this PR. Likewise the functions that are `ensureX` in other stores serve an important purpose.

You should be making this behavior _standard_ rather than removing it.

Reply:
an optional `force` param has been added back to the project store's `fetch` function. The default behavior is to use cached values when applicable unless `force` is set to `true` (which we are not using yet).